File: ThreadLab2C#/MinimumFinder/ArrayMinimumFinderWithCountDown.cs
namespace ConsoleApp1
{
    using System;
    using System.Diagnostics;
    using System.Threading;
    using System.Threading.Tasks;

    public class ArrayMinimumFinderWithCountDown
    {
        private readonly ArrayMinimumManager _manager;
        private readonly int _threadCount;

        public ArrayMinimumFinderWithCountDown(ArrayMinimumManager manager, int threadCount)
        {
            _manager = manager;
            _threadCount = threadCount;
        }

        public void FindMinInArr(int[] arr, int start, int end)
        {
            int min = arr[start];
            for (int i = start + 1; i < end; i++)
            {
                if (arr[i] < min)
                {
                    min = arr[i];
                }
            }

            _manager.UpdateMin(min);
        }

        public void StartFindingMin(int[] arr)
        {
            int size = arr.Length;

            var sw = Stopwatch.StartNew();

            CountdownEvent countdown = new CountdownEvent(_threadCount);

            for (int i = 0; i < _threadCount; i++)
            {
                var segment = GetSegment(i, _threadCount, size);
                Task.Run(() => ThreadWorker(arr, segment.Start, segment.End, countdown));
            }

            countdown.Wait();


            sw.Stop();
            Console.WriteLine($"----Search time with Tasks: {sw.ElapsedMilliseconds} ms");
        }

        private void ThreadWorker(int[] arr, int start, int end, CountdownEvent countdown)
        {
            try
            {
                FindMinInArr(arr, start, end);
            }
            finally
            {
                countdown.Signal();
            }
        }

        private static Bound GetSegment(int index, int segmentsCount, int length)
        {
            int segmentSize = length / segmentsCount;
            int start = index * segmentSize;
            int end = (index == segmentsCount - 1) ? length : start + segmentSize;
            return new Bound(start, end);
        }
    }
}

File: ThreadLab2Java/Threads/src/ArrayMinimumFinderWithCountDown.java
import java.util.concurrent.CountDownLatch;

public class ArrayMinimumFinderWithCountDown {
    private final ArrayMinimumManager manager;
    private final int threadCount;

    public ArrayMinimumFinderWithCountDown(ArrayMinimumManager manager, int threadCount) {
        this.manager = manager;
        this.threadCount = threadCount;
    }

    public void findMinInArr(int[] arr, int start, int end) {
        int min = arr[start];
        for (int i = start + 1; i < end; i++) {
            if (arr[i] < min) {
                min = arr[i];
            }
        }

        manager.updateMin(min);
    }

    public void startFindingMin(int[] arr) {
        int size = arr.length;

        long startTime = System.currentTimeMillis();

        CountDownLatch countdown = new CountDownLatch(threadCount);

        for (int i = 0; i < threadCount; i++) {
            Bound segment = getSegment(i, threadCount, size);
            new Thread(() -> threadWorker(arr, segment.start(), segment.end(), countdown)).start();
        }

        try {
            countdown.await();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }

        long endTime = System.currentTimeMillis();
        System.out.println("---Search time with CountDownLatch: " + (endTime - startTime) + " ms");
    }

    private void threadWorker(int[] arr, int start, int end, CountDownLatch countdown) {
        try {
            findMinInArr(arr, start, end);
        } finally {
            countdown.countDown();
        }
    }

    private static Bound getSegment(int i, int arrCount, int length) {
        int segmentLength = length / arrCount;

        int firstIndex = segmentLength * i;
        int lastIndex = segmentLength * (i + 1);

        if (lastIndex > length) {
            lastIndex = length;
        }
    
        return new Bound(firstIndex, lastIndex);
    }
}